# Data

Data yang digunakan bersumber dari Project Gutenberg, yaitu novel Alice In Wonderland (Chapter 1)

## Import Data

In [38]:
with open('data/AliceInWonderland_Ch01.txt', 'r', encoding = 'utf-8') as f:
    docs = f.read()

In [39]:
# total karakter
print("Total karakter:", len(docs))

Total karakter: 11385


# Token Awal

In [40]:
from nltk.tokenize import word_tokenize

token_awal = word_tokenize(docs)
# banyak token awal (sebelum praproses)
print("Token awal:", len(token_awal))

Token awal: 2615


# Praproses

## Remove Punctuation

In [41]:
import string

In [42]:
# hapus titik, koma, dan tanda baca lainnya
new_docs = docs.translate(str.maketrans('', '', string.punctuation))

print(new_docs[:5000])

Alice was beginning to get very tired of sitting by her sister on the
bank and of having nothing to do once or twice she had peeped into
the book her sister was reading but it had no pictures or
conversations in it “and what is the use of a book” thought Alice
“without pictures or conversations”

So she was considering in her own mind as well as she could for the
hot day made her feel very sleepy and stupid whether the pleasure of
making a daisychain would be worth the trouble of getting up and
picking the daisies when suddenly a White Rabbit with pink eyes ran
close by her

There was nothing so very remarkable in that nor did Alice think it
so very much out of the way to hear the Rabbit say to itself “Oh
dear Oh dear I shall be late” when she thought it over afterwards
it occurred to her that she ought to have wondered at this but at the
time it all seemed quite natural but when the Rabbit actually took a
watch out of its waistcoatpocket and looked at it and then hurried
on Alice star

“without pictures or conversations” dsb, tidak ikut dibersihkan karena bukan ASCII 

In [43]:
import re

In [44]:
# hapus tanda baca yang lain 
# termasuk yang bukan ASCII
new_docs = re.sub(r'[^\w\s]', '', new_docs)
print(new_docs[:5000])

Alice was beginning to get very tired of sitting by her sister on the
bank and of having nothing to do once or twice she had peeped into
the book her sister was reading but it had no pictures or
conversations in it and what is the use of a book thought Alice
without pictures or conversations

So she was considering in her own mind as well as she could for the
hot day made her feel very sleepy and stupid whether the pleasure of
making a daisychain would be worth the trouble of getting up and
picking the daisies when suddenly a White Rabbit with pink eyes ran
close by her

There was nothing so very remarkable in that nor did Alice think it
so very much out of the way to hear the Rabbit say to itself Oh
dear Oh dear I shall be late when she thought it over afterwards
it occurred to her that she ought to have wondered at this but at the
time it all seemed quite natural but when the Rabbit actually took a
watch out of its waistcoatpocket and looked at it and then hurried
on Alice started to

## Stopwords

Kata yang berdiri sendiri dan tidak ada maknanya, seperti kata hubung

In [45]:
import nltk
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('punkt_tab')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\LENOVO\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\LENOVO\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\LENOVO\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [46]:
# import english stopwords
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

# gunakan stopwords bahasa inggris
stop_words = set(stopwords.words('english'))

# tokenisasi
word_tokens = word_tokenize(new_docs)

In [47]:
# apply stopword 
filtered_docs = [word for word in word_tokens if not word.lower() in stop_words]

In [48]:
# print
print(" ".join(filtered_docs[:200]))

Alice beginning get tired sitting sister bank nothing twice peeped book sister reading pictures conversations use book thought Alice without pictures conversations considering mind well could hot day made feel sleepy stupid whether pleasure making daisychain would worth trouble getting picking daisies suddenly White Rabbit pink eyes ran close nothing remarkable Alice think much way hear Rabbit say Oh dear Oh dear shall late thought afterwards occurred ought wondered time seemed quite natural Rabbit actually took watch waistcoatpocket looked hurried Alice started feet flashed across mind never seen rabbit either waistcoatpocket watch take burning curiosity ran across field fortunately time see pop large rabbithole hedge another moment went Alice never considering world get rabbithole went straight like tunnel way dipped suddenly suddenly Alice moment think stopping found falling deep well Either well deep fell slowly plenty time went look wonder going happen next First tried look make c

Kata seperti 'was' dihapus

## Lowercase

In [49]:
# mengubaah teks di filtered_docs menjadi lower semua
# karena bentuknya list maka gunakan list comprehension
lowered_docs = [word.lower() for word in filtered_docs]
print(lowered_docs[:5000])

['alice', 'beginning', 'get', 'tired', 'sitting', 'sister', 'bank', 'nothing', 'twice', 'peeped', 'book', 'sister', 'reading', 'pictures', 'conversations', 'use', 'book', 'thought', 'alice', 'without', 'pictures', 'conversations', 'considering', 'mind', 'well', 'could', 'hot', 'day', 'made', 'feel', 'sleepy', 'stupid', 'whether', 'pleasure', 'making', 'daisychain', 'would', 'worth', 'trouble', 'getting', 'picking', 'daisies', 'suddenly', 'white', 'rabbit', 'pink', 'eyes', 'ran', 'close', 'nothing', 'remarkable', 'alice', 'think', 'much', 'way', 'hear', 'rabbit', 'say', 'oh', 'dear', 'oh', 'dear', 'shall', 'late', 'thought', 'afterwards', 'occurred', 'ought', 'wondered', 'time', 'seemed', 'quite', 'natural', 'rabbit', 'actually', 'took', 'watch', 'waistcoatpocket', 'looked', 'hurried', 'alice', 'started', 'feet', 'flashed', 'across', 'mind', 'never', 'seen', 'rabbit', 'either', 'waistcoatpocket', 'watch', 'take', 'burning', 'curiosity', 'ran', 'across', 'field', 'fortunately', 'time', '

## Stemming

Ternyata stemming tidak memberikan hasil yang baik, jadi tidak memilih stemming untuk dibagian praproses

## Lemmatization

In [50]:
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\LENOVO\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\LENOVO\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


True

In [51]:
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize

wnl = WordNetLemmatizer()

In [52]:
# karena terpisah perkata, gabungkan dulu
words = word_tokenize(" ".join(lowered_docs))

new_text = " ".join([wnl.lemmatize(w) for w in words])
print(new_text[:500])

alice beginning get tired sitting sister bank nothing twice peeped book sister reading picture conversation use book thought alice without picture conversation considering mind well could hot day made feel sleepy stupid whether pleasure making daisychain would worth trouble getting picking daisy suddenly white rabbit pink eye ran close nothing remarkable alice think much way hear rabbit say oh dear oh dear shall late thought afterwards occurred ought wondered time seemed quite natural rabbit act


# Tokenisasi

In [53]:
# tokenisasi
data_tokenized = word_tokenize(" ".join(lowered_docs))
print(data_tokenized[:500])

['alice', 'beginning', 'get', 'tired', 'sitting', 'sister', 'bank', 'nothing', 'twice', 'peeped', 'book', 'sister', 'reading', 'pictures', 'conversations', 'use', 'book', 'thought', 'alice', 'without', 'pictures', 'conversations', 'considering', 'mind', 'well', 'could', 'hot', 'day', 'made', 'feel', 'sleepy', 'stupid', 'whether', 'pleasure', 'making', 'daisychain', 'would', 'worth', 'trouble', 'getting', 'picking', 'daisies', 'suddenly', 'white', 'rabbit', 'pink', 'eyes', 'ran', 'close', 'nothing', 'remarkable', 'alice', 'think', 'much', 'way', 'hear', 'rabbit', 'say', 'oh', 'dear', 'oh', 'dear', 'shall', 'late', 'thought', 'afterwards', 'occurred', 'ought', 'wondered', 'time', 'seemed', 'quite', 'natural', 'rabbit', 'actually', 'took', 'watch', 'waistcoatpocket', 'looked', 'hurried', 'alice', 'started', 'feet', 'flashed', 'across', 'mind', 'never', 'seen', 'rabbit', 'either', 'waistcoatpocket', 'watch', 'take', 'burning', 'curiosity', 'ran', 'across', 'field', 'fortunately', 'time', '

## Token Akhir

In [54]:
# melihat banyak token terakhir setelah praproses
len(data_tokenized)

977

Dari total token awal sebanyak 2615 tersisa 977 token setelah dilakukan praproses

# Text Vectorization

In [55]:
# karena token cukup banyak, ambil 500 token pertama saja
data_tokenized = data_tokenized[:500]
print(data_tokenized)

['alice', 'beginning', 'get', 'tired', 'sitting', 'sister', 'bank', 'nothing', 'twice', 'peeped', 'book', 'sister', 'reading', 'pictures', 'conversations', 'use', 'book', 'thought', 'alice', 'without', 'pictures', 'conversations', 'considering', 'mind', 'well', 'could', 'hot', 'day', 'made', 'feel', 'sleepy', 'stupid', 'whether', 'pleasure', 'making', 'daisychain', 'would', 'worth', 'trouble', 'getting', 'picking', 'daisies', 'suddenly', 'white', 'rabbit', 'pink', 'eyes', 'ran', 'close', 'nothing', 'remarkable', 'alice', 'think', 'much', 'way', 'hear', 'rabbit', 'say', 'oh', 'dear', 'oh', 'dear', 'shall', 'late', 'thought', 'afterwards', 'occurred', 'ought', 'wondered', 'time', 'seemed', 'quite', 'natural', 'rabbit', 'actually', 'took', 'watch', 'waistcoatpocket', 'looked', 'hurried', 'alice', 'started', 'feet', 'flashed', 'across', 'mind', 'never', 'seen', 'rabbit', 'either', 'waistcoatpocket', 'watch', 'take', 'burning', 'curiosity', 'ran', 'across', 'field', 'fortunately', 'time', '

## Bag Of Words

In [56]:
from sklearn.feature_extraction.text import CountVectorizer

In [57]:
vectorizer = CountVectorizer()
# input harus berupa string, jadi gabungkan token kembali
X = vectorizer.fit_transform([" ".join(data_tokenized)])

print(X.toarray())
print(vectorizer.get_feature_names_out())

[[ 2  1  1  1  2 14  1  1  2  1  1  2  2  1  1  1  1  2  4  3  1  1  1  1
   2  1  1  1  2  1  1  4  1  2  2  1  2  2  2  2  1  1  2  1  1  1  1  1
   2  1  3  2  2  4  1  1  1  1  1  1  1  1  1  1  1  1  1  1  2  5  4  1
   1  1  1  2  1  1  4  1  2  1  1  2  3  1  1  1  1  1  1  2  1  1  4  3
   1  1  1  2  1  1  1  2  2  1  1  1  1  2  1  1  1  1  1  1  1  1  1  1
   1  1  2  2  2  1  1  2  1  1  1  1  2  2  1  1  1  1  4  1  1  1  1  2
   1  2  1  2  2  3  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  2  1  2
   1  4  1  3  1  1  1  1  4  1  1  1  4  1  1  3  5  1  1  1  1  1  1  1
   1  1  1  1  1  3  1  1  1  1  1  1  1  2  1  1  6  2  2  2  1  1  1  3
   1  1  1  1  1  1  1  5  2  1  6  1  1  2  1  5  1  1  1  1  1  1  2  1
   2  1  1  1  2  1  2  1  1  1  1  1  2  1  1  1  5  1  1  1  1  2  2  1
   9  1  4  1  1  2  6  1  1  2  1  2  1  1  1  1  1  1  2  1  1  3  1  2
   1  1  1  2  5  5  5  1  1  2  1  1  1  5  1  1  1  1  1  1  3  1  1  1
   1  1]]
['across' 'actually' 'afraid

## TF-IDF

In [58]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [59]:
vectorizer = TfidfVectorizer()
# input harus berupa string, jadi gabungkan token kembali
X = vectorizer.fit_transform([" ".join(data_tokenized)])

print(X.toarray())
print(vectorizer.get_feature_names_out())

[[0.05463584 0.02731792 0.02731792 0.02731792 0.05463584 0.38245086
  0.02731792 0.02731792 0.05463584 0.02731792 0.02731792 0.05463584
  0.05463584 0.02731792 0.02731792 0.02731792 0.02731792 0.05463584
  0.10927167 0.08195375 0.02731792 0.02731792 0.02731792 0.02731792
  0.05463584 0.02731792 0.02731792 0.02731792 0.05463584 0.02731792
  0.02731792 0.10927167 0.02731792 0.05463584 0.05463584 0.02731792
  0.05463584 0.05463584 0.05463584 0.05463584 0.02731792 0.02731792
  0.05463584 0.02731792 0.02731792 0.02731792 0.02731792 0.02731792
  0.05463584 0.02731792 0.08195375 0.05463584 0.05463584 0.10927167
  0.02731792 0.02731792 0.02731792 0.02731792 0.02731792 0.02731792
  0.02731792 0.02731792 0.02731792 0.02731792 0.02731792 0.02731792
  0.02731792 0.02731792 0.05463584 0.13658959 0.10927167 0.02731792
  0.02731792 0.02731792 0.02731792 0.05463584 0.02731792 0.02731792
  0.10927167 0.02731792 0.05463584 0.02731792 0.02731792 0.05463584
  0.08195375 0.02731792 0.02731792 0.02731792 0.